### DISCUSSION TITLE
Error: Text content does not match server-rendered HTML.

### DISCUSSION THREAD

**Discussion by MikevPeeren** [vercel/next.js#41319] [Help]
I am using i18n on my pages. The pages have no form of Date or Timestring, however this error keeps popping up on the login page. I am using react-i18next does anyone know how to fix this ?

**Comment by icyJoseph** [comment]
Can you tell where in the DOM the divergence happens? In production it can be a bit tricky, but possible by placing a debugger on the warning and then stepping through to see what it expected and what it received.

I don't want to assume which library could be triggering this until we can identify the DOM element that diverges. Making a request with disabled JS (or from a non-browser env) might also help out to identify the offending node.

**Reply by MikevPeeren** [reply]
I have found the error. I am initialising the default language in my i18n as following:

`lng: getDefaultLanguage(),`

In my getDefaultLanguage i am getting the stored language like so `local.getItem("i18nextLng")`. This can be nl-NL for example but everything on the server gets built as english which is why the error is present. No idea how to solve this tho.

**Reply by icyJoseph** [reply]
Ah, I am a bit short on time, now but perhaps you can use the `Accept-Language` header to try and pre-render the page as the user desires it? 

Or perhaps you should save the user preference as cookie instead. Maybe this talk on dark mode perfectionism can help you out: https://www.youtube.com/watch?v=UhLWAQdsgBY  - https://github.com/NDAify/ndaify-web/blob/132343839e0d38a0410f9da014543619849ff600/src/lib/useTheme.js#L38

**Reply by MikevPeeren** [reply]
That would still only render English on the server no ? The i18nextLng is taken from localStorage how would a Cookie be better ?

**Reply by icyJoseph** [reply]
The cookie can travel with the page request, and you could read on the server what was set by the user. LocalStorage doesn't travel, so to speak. 

In their case they use `getInitialProps` to get access to the incoming request, and do: 


```js
 const preferredTheme = Component.themeOverride || getThemePreference(ctx);
```

You'd do something like:

```js
 const preferredLanguage =  getLanguagePreference(ctx)  || getDefaultLanguage();
```

Lots of pseudo-code and ideas only though, but that's roughly the architecture that should fix this.

**Reply by MikevPeeren** [reply]
Still can't seem to fix it with a cookie, maybe you can help me further if you see my code. Also have no idea how to implement your above solution.

i18n setup is as following:
```
i18n
  .use(LanguageDetector)
  .use(initReactI18next)
  .init({
    // lng: getDefaultLanguage(),
    fallbackLng: {
      "en-GB": ["en-GB"],
      "nl-NL": ["nl-NL", "en-GB"],
      "nl-BE": ["nl-BE", "nl-NL", "en-GB"],
      default: ["en-GB"],
    },
    // eslint-disable-next-line no-undef
    debug: true,
    resources: {
      "en-GB": en,
      "nl-NL": nl,
      // eslint-disable-next-line camelcase
      "nl-BE": nl_BE,
    },
    interpolation: {
      escapeValue: false, // not needed for react as it escapes by default
    },
    detection: {
      order: ["cookie", "navigator"],
      caches: ["cookie"],
      lookupFromPathIndex: 0,
    },
  });
```

Error with a cookie set to something different than what the server expects.
<img width="1681" alt="Screenshot 2022-10-13 at 13 37 40" src="https://user-images.githubusercontent.com/15649037/195586690-110033cf-2b8a-4414-aac2-b088c269432a.png">

**Reply by icyJoseph** [reply]
Sorry I am a bit short on time this week, but apparently, https://nextjs.org/docs/advanced-features/i18n-routing#leveraging-the-next_locale-cookie there's a solution for this, you need to set the `NEXT_LOCALE=the-locale` cookie, which is a JS available cookie.

Are you using:

-  https://github.com/i18next/next-i18next or,
-  the base package: https://www.i18next.com/overview/supported-frameworks, 
-  the "generic" react version, https://react.i18next.com/ this one still support SSR, but you gotta do a bit of extra work 

I'd say, take the first package, which is based of i18next anyway. The "next" word does make this confusing 😅

**Reply by MikevPeeren** [reply]
NP on the short time on your part you ar ehelping allot alteady so far so thanks for that. I am using react-i18next but we have the transition to next-i18next planned for the future however that is not feasable for now.

**Reply by MikevPeeren** [reply]
I am also wondering if I change the language in a useEffect if that would also solve the problem rather than at the i18n init.

**Reply by Enache-Ionut** [reply]
@MikevPeeren do you have any update on this? Unfortunately, I encountered the same issue.

**Reply by MikevPeeren** [reply]
The problem I had was that I initialising the language server side and also checking for this client side ( localStorage & fallback ). I moved that logic to a useEffect which made most of our errors decline, We still have them however these are likely caused by other effects.

So my last comment on Oct 15 did actually resolve this for me.

**Comment by hyjoong** [comment]
<img width="968" alt="image" src="https://user-images.githubusercontent.com/70426440/214007175-ceca3e17-fd51-48ef-ae05-0252e35d1420.png">

I encountered the same issue and solved it in the following way.

import useTranslation from `next-i18next` not `react-i18next`

 
```diff
- import { useTranslation } from "react-i18next";
+ import { useTranslation } from "next-i18next";
```

**Reply by oussj** [reply]
Thanks a lot, I wrote a small eslint custom rule to avoid this : 
```
    "no-restricted-imports": [
      "error",
      {
        "paths": [
          {
            "name": "react-i18next",
            "message": "Please use next-i18next import instead to avoid hydration errors"
          }
        ]
      }
    ],
```

**Reply by andriivam** [reply]
So funny for me it opposite. When i changed import to react all errors are gone)

**Reply by Mahmoud-Anani** [reply]
هذا الحل يجعلها تعمل في بعض الوقت و البعض الاخر يرجع الخطأ عادي.

**Reply by SakarDev** [reply]
> So funny for me it opposite. When i changed import to react all errors are gone)

Same!
